src/bin/krillup.rs

```diff
@@ -1,82 +1,145 @@
 extern crate krill;
 
-use clap::{App, Arg};
-use log::info;
+use clap::{App, Arg, ArgMatches, SubCommand};
+use log::{info, LevelFilter};
 
 use krill::{
     constants::{KRILL_DEFAULT_CONFIG_FILE, KRILL_UP_APP, KRILL_VERSION},
-    daemon::{config::Config, properties::PropertiesManager},
-    upgrades::{prepare_upgrade_data_migrations, UpgradeMode},
+    daemon::{
+        config::{Config, LogType},
+        properties::PropertiesManager,
+    },
+    upgrades::{data_migration::migrate, prepare_upgrade_data_migrations, UpgradeMode},
 };
+use url::Url;
 
-#[tokio::main]
-async fn main() {
-    let matches = App::new(KRILL_UP_APP)
-        .version(KRILL_VERSION)
-        .about("\nThis tool can be used to reduce the risk and time needed for Krill upgrades, by preparing and verifying any data migrations that would be needed. The data_dir setting from the provided configuration file is used to find the data to migrate, and prepared data will be saved under 'data_dir/upgrade-data'.")
-        .arg(
-            Arg::with_name("config")
-                .short("c")
-                .long("config")
-                .value_name("FILE")
-                .help(&format!(
-                    "Override the path to the config file (default: '{}')",
-                    KRILL_DEFAULT_CONFIG_FILE
-                ))
-                .required(false),
-        )
-        .get_matches();
-
-    let config_file = matches.value_of("config").unwrap_or(KRILL_DEFAULT_CONFIG_FILE);
-
-    match Config::create(config_file, true) {
-        Ok(config) => {
-            let properties_manager = match PropertiesManager::create(&config.storage_uri, config.use_history_cache)
-            {
-                Ok(mgr) => mgr,
-                Err(e) => {
-                    eprintln!("*** Error Preparing Data Migration ***");
-                    eprintln!("{}", e);
-                    eprintln!();
-                    eprintln!("Note that your server data has NOT been modified. Do not upgrade krill itself yet!");
-                    eprintln!("If you did upgrade krill, then downgrade it to your previous installed version.");
-                    ::std::process::exit(1);
-                }
-            };
+fn main() {
+    let matches = make_matches();
+
+    match parse_matches(matches) {
```

Comment:
This flow feels illogical. I am down the file and then continuing up here again. Then again this is probably a personal preference so feel to ignore it :-)

Reply:
I usually like to have `main` as the first thing. But the whole match could be in a separate function as well. It's personal preference territory.